runtime/major_gc.c

```diff
@@ -398,6 +393,13 @@ static void mark_slice (intnat work)
       CAMLassert (start == 0);
       v = *--gray_vals_ptr;
       CAMLassert (Is_gray_val (v));
+#ifdef NO_NAKED_POINTERS
+      if (Tag_val(v) == Closure_tag) {
+        /* Skip the code pointers and integers at beginning of closure;
+           start scanning at the first word of the environment part. */
+        start = Start_env_closinfo(Closinfo_val(v));
```

Comment:
Maybe we could have an assertion that `Start_env_closinfo` does indeed lie within the value `v`?

Reply:
All right.  See commit c7a226436 .